Add ability to include image pull policy with runtime image

akchinSTC opened the pull request:
Fixes #1233



 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

kevin-bates commented on elyra/pipeline/pipeline.py:
```diff
@@ -141,6 +141,14 @@ def env_vars_as_dict(self, logger: Optional[object] = None) -> Dict:
+    @property
+    def image_pull_policy(self):
+        image_tag = self._runtime_image.split(":")[1]
+        if image_tag == "latest":
```
Since 'latest' is also implied when a tag does not exist, I believe that should be handled as well.

akchinSTC replied:
Updated

kevin-bates reviewed (commented):
I understood the issue to be that we need a way for a user to specify the pull policy of a runtime image.  This seems to only retain the k8s default behavior.  

Use case:  I'm developing a new image, tagged with 'dev', but I want that image to use a pull policy of 'Always', how can I do that?

I figured this would require an enum added to the metadata of ['Always', 'Never' and 'IfNotPresent'] or something like that.  Then, the runtime would consult the instance and set the desired pull policy.

Hmm - thinking about that implies that we must also convey the runtime-image metadata name (or have a way to get it), but I believe we only set the image name in the pipeline props, so I suppose that would be an issue.  I suppose pull policy could come from the client (front end) as well.

akchinSTC commented:
Based on the initial comments in the issue, it seemed like we wanted a behavior set based on image name rather than adding more configuration options in the metadata at the image level or common properties at the node level. 

thoughts @lresende ?

lresende commented:
My understanding is:
- Runtime-image schema needs to be updated with an optional image_policy enum field
- When a user is creating/editing a runtime-image, he can select the associated image policy
   - As @kevin-bates mentioned (and assuming version tags are immutable), the pull policy only makes sense with non-version tags such as none, latest, dev, etc. Having said that, I don't think we need to make any attempt to add any logic to the selection, either the user selects one or not independent of the type of tag associated with the image. 
- As we did with resources, if the user does not select an image_policy value, we should not send anything and the K8s default behavior should apply

Does this clarify a bit more?

kevin-bates commented:
Yes. Thanks. Will we just copy the pull policy and store the value, if present, in the node properties as we do today with the image name, or change to using a by-reference approach for the runtime instance? If the former, updates to the image name should unconditionally reset the pull policy value.

lresende commented:
I would say similar to what we do today, where we would probably add an additional field ... something like `runtime-image-pull-policy`

akchinSTC commented:
I think I opted for the latter approach from Kevin. The schema has been updated and I do a lookup of the image metadata when we generate the pipeline to find the pull policy and set it if it exists. I modified the get_runtime_configuration function to do double duty and get both runtime and image metadata depending on input, but it might be a little messy since I think they return different types. I may just create another function for either.

kevin-bates commented:
I think this approach may be better than by-value (where the pull-policy is persisted) for the following reasons:
1. The pull policy isn't persisted in the pipeline file beside the image.
2. If the server doesn't find a matching image name in its set of runtime-images, the Kubernetes behavior kicks in, and that seems sufficient.  Users can create a matching runtime-image instance associated with the referenced image name if they want pull behavior other than the defaults.

Since pull policy is really an attribute of the processing server and not so much an attribute of the image, this approach satisfies that.  It also prevents the "stickiness" of existing pipelines and folks having to potentially visit each node to change the pull policy.

kevin-bates reviewed (approved):
Thanks Alan.

lresende commented on elyra/metadata/schemas/runtime-image.json:
```diff
@@ -37,6 +37,12 @@
+          "title": "Image Pull Policy",
+          "description": "The pull policy to use when selecting this image",
+          "type": "string",
+          "enum": ["Always", "IfNotPresent", "Never"]
```
```suggestion
          "enum": ["Always", "IfNotPresent", "Never"],
          "uihints": {
            "field_type": "dropdown"
          }          
```

lresende reviewed (commented):
LGTM, I added a suggestion to make the field a dropdown. We also need to resolve conflicts

akchinSTC commented:
> LGTM, I added a suggestion to make the field a dropdown. We also need to resolve conflicts

Updated and rebased on master after airflow feature merge. 
Airflow template and processor have been updated with the pull policy logic.
